src/main/scala/io/iohk/ethereum/ledger/Ledger.scala

```diff
@@ -44,32 +45,41 @@ object Ledger extends Logger {
     stateStorage: NodeStorage):
   ExecResult = {
     val blockchain = BlockchainImpl(storages)
-    val initialWorldStateProxy = InMemoryWorldStateProxy(storages, stateStorage,
+    val initialWorld = InMemoryWorldStateProxy(storages, stateStorage,
       blockchain.getBlockHeaderByHash(block.header.hash).map(_.stateRoot)
     )
-    block.body.transactionList.foldLeft[ExecResult](initialWorldStateProxy -> 0) {
-      case ((worldStateProxy, acumGas), stx) =>
-        val result: Either[String, PR] = for {
-          _ <- validateTransaction(stx, worldStateProxy, 0, block)
-          worldStateProxy1 = updateAccountBeforeExecution(stx, worldStateProxy)
-          result <- execute(stx, block.header, worldStateProxy1)
-        } yield result
-
-        result match {
-          case Right(theResult) =>
-            val gasUsed = stx.tx.gasLimit - theResult.gasRemaining
-            val refundGasFn = {
-              if (theResult.gasRemaining > 0) (refundGas _).curried(Address(block.header.beneficiary))(gasUsed)
-              else identity[InMemoryWorldStateProxy] _
-            }
-            val payBeneficiariesFn = (payForGasUsedToBeneficiary _).curried(stx)
-            val deleteAccountsFn = (deleteAccounts _).curried(theResult.addressesToDelete)
-            (refundGasFn andThen payBeneficiariesFn andThen deleteAccountsFn) (theResult.world) -> (gasUsed + acumGas)
-          case Left(error) => throw new RuntimeException(error)
+    block.body.transactionList.foldLeft[ExecResult](initialWorld -> 0) {
+      case ((world, acumGas), stx) =>
+        //TODO: change 0 -> acumGas ?
+        validateTransaction(stx, world, 0, block.header) match {
+          case Left(err) =>
+            throw new RuntimeException(err)
+
+          case Right(_) =>
+            val (resultWorld, gasUsed) = executeTransaction(stx, block.header, world)
+            (resultWorld, acumGas + gasUsed)
         }
     }
   }
 
+  private[ledger] def executeTransaction(stx: SignedTransaction, blockHeader: BlockHeader, world: InMemoryWorldStateProxy): ExecResult = {
+    val gasPrice = UInt256(stx.tx.gasPrice)
+    val gasLimit = UInt256(stx.tx.gasLimit)
+
+    val world1 = updateSenderAccountBeforeExecution(stx, world)
+    val result = runVM(stx, blockHeader, world1)
+
+    val gasUsed = if(result.error.isDefined) gasLimit else gasLimit - result.gasRemaining
+    val gasRefund = calcGasRefund(stx, result)
+
+    val refundGasFn = pay(stx.recoveredSenderAddress.get, gasRefund * gasPrice) _
+    val payMinerForGasFn = pay(Address(blockHeader.beneficiary), (gasLimit - gasRefund) * gasPrice) _
+    val deleteAccountsFn = deleteAccounts(result.addressesToDelete) _
+
+    val world2 = (refundGasFn andThen payMinerForGasFn andThen deleteAccountsFn) (result.world)
```

Comment:
Is it okay for accounts to be deleted in case the execution resulted in an error? At least `EthereumJ` seems to clear the list of accounts to be deleted when an exception ocurrs.